File: Senior answers/SC2079-MDP-main-main/android/app/src/main/java/com/mdp26/mdp20/BluetoothActivity.java
package com.mdp26.mdp20;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AlertDialog;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.Toolbar;
import androidx.appcompat.widget.SwitchCompat;
import android.widget.LinearLayout;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import android.Manifest;
import android.annotation.SuppressLint;
import android.app.Activity;
import android.bluetooth.BluetoothAdapter;
import android.bluetooth.BluetoothDevice;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.content.pm.PackageManager;
import android.location.LocationManager;
import android.media.MediaPlayer;
import android.net.Uri;
import android.os.Bundle;
import android.provider.Settings;
import android.util.Log;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;
import android.widget.Toast;

import com.mdp26.mdp20.bluetooth.BluetoothConnection;
import com.mdp26.mdp20.bluetooth.BluetoothDeviceAdapter;
import com.mdp26.mdp20.bluetooth.BluetoothInfoReceiver;
import com.mdp26.mdp20.bluetooth.BluetoothMessage;
import com.mdp26.mdp20.bluetooth.BluetoothMessageParser;
import com.mdp26.mdp20.bluetooth.BluetoothMessageReceiver;

import java.util.ArrayList;

public class BluetoothActivity extends AppCompatActivity {

    private static final String TAG = "BluetoothActivity";
    private static final int BLUETOOTH_PERMISSIONS_REQUEST_CODE = 96;
    private static final int DISCOVERABLE_DURATION = 300; 
    private MyApplication myApp; 
    private BroadcastReceiver infoReceiver; 
    private BroadcastReceiver msgReceiver; 
    private BluetoothDeviceAdapter bluetoothDeviceAdapter; 

    private ActivityResultLauncher<Intent> requestEnableBluetooth;
    private ActivityResultLauncher<Intent> requestDiscoverable;

    private TextView receivedMsgView;
    private LinearLayout connectedPanel;
    private TextView connectedText;
    private View indicatorCircle;
    private SwitchCompat discoverSwitch;
    private View saberBlade;
    private android.animation.ValueAnimator saberAnimator;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_bluetooth);

        Toolbar toolbar = findViewById(R.id.topAppBar);
        setSupportActionBar(toolbar);

        myApp = (MyApplication) getApplication();

        String[] permissions = new String[] {
                Manifest.permission.BLUETOOTH_SCAN,
                Manifest.permission.BLUETOOTH_CONNECT,
                Manifest.permission.BLUETOOTH_ADVERTISE,
                Manifest.permission.ACCESS_FINE_LOCATION, 
                Manifest.permission.ACCESS_COARSE_LOCATION, 
        };
        requestPermissions(permissions, BLUETOOTH_PERMISSIONS_REQUEST_CODE);

        bindUI();

        requestEnableBluetooth = registerForActivityResult(
                new ActivityResultContracts.StartActivityForResult(),
                result -> {
                    if (result.getResultCode() == Activity.RESULT_OK) {
                        Log.d(TAG, "Bluetooth enabled.");
                        bootBluetoothService();
                    }
                });

        requestDiscoverable = registerForActivityResult(
                new ActivityResultContracts.StartActivityForResult(),
                result -> {
                    
                    if (result.getResultCode() == DISCOVERABLE_DURATION
                            || result.getResultCode() == Activity.RESULT_OK) {
                        Log.d(TAG, "Bluetooth Discovery on for " + DISCOVERABLE_DURATION + "s");
                        
                        if (discoverSwitch != null)
                            discoverSwitch.setChecked(true);

                        animateLightsaber(true);

                        new android.os.Handler(android.os.Looper.getMainLooper()).postDelayed(() -> {
                            if (discoverSwitch != null) {
                                discoverSwitch.setChecked(false);
                                animateLightsaber(false); 
                                Toast.makeText(this, "Discoverable mode disabled", Toast.LENGTH_SHORT).show();
                            }
                        }, DISCOVERABLE_DURATION * 1000L);
                    } else {
                        
                        Log.d(TAG, "Bluetooth Discovery cancelled/failed");
                        if (discoverSwitch != null)
                            discoverSwitch.setChecked(false);
                        animateLightsaber(false); 
                    }
                });

        infoReceiver = new BluetoothInfoReceiver(this::onBluetoothInfoReceived);
        for (IntentFilter intentFilter : BluetoothInfoReceiver.DEFAULT_FILTERS) {
            
            getApplicationContext().registerReceiver(infoReceiver, intentFilter, RECEIVER_EXPORTED);
        }

        msgReceiver = new BluetoothMessageReceiver(BluetoothMessageParser.ofDefault(), this::onMsgReceived);
        getApplicationContext().registerReceiver(msgReceiver,
                new IntentFilter(BluetoothMessageReceiver.ACTION_MSG_READ), RECEIVER_NOT_EXPORTED);

        LocationManager locationManager = (LocationManager) getSystemService(Context.LOCATION_SERVICE);
        boolean isGpsEnabled = locationManager.isProviderEnabled(LocationManager.GPS_PROVIDER);
        if (!isGpsEnabled) {
            Log.d(TAG, "GPS / Location is not on, won't be able to discover non-paired devices.");
            Toast.makeText(this, "Turn on Location to Scan for devices.", Toast.LENGTH_SHORT).show();
        }
    }

    private void bindUI() {
        
        RecyclerView recyclerView = findViewById(R.id.btDeviceList);
        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        bluetoothDeviceAdapter = new BluetoothDeviceAdapter(this, new ArrayList<>(), device -> {
            Toast.makeText(this, "Connecting to " + device.name(), Toast.LENGTH_SHORT).show();
            myApp.btInterface().connectAsClient(device.btDevice());
        });
        recyclerView.setAdapter(bluetoothDeviceAdapter);
        findViewById(R.id.btnScan).setOnClickListener(view -> reloadDiscoveredDevices());

        discoverSwitch = findViewById(R.id.switchDiscoverable);
        discoverSwitch.setOnCheckedChangeListener((buttonView, isChecked) -> {
            
            if (buttonView.isPressed() && isChecked) {
                enableDeviceDiscovery();
            } else if (buttonView.isPressed() && !isChecked) {
                
                animateLightsaber(false);
            }
        });

        android.widget.ImageButton themeToggle = findViewById(R.id.btnThemeToggle);
        themeToggle.setOnClickListener(v -> toggleTheme());

        Button canvasButton = findViewById(R.id.btnCanvas);
        canvasButton.setVisibility(View.VISIBLE); 
        canvasButton.setOnClickListener(view -> startActivity(new Intent(this, CanvasActivity.class)));

        findViewById(R.id.btnBigRed).setOnClickListener(v -> {
            startActivity(new Intent(BluetoothActivity.this, HyperspaceActivity.class));
        });

        connectedPanel = findViewById(R.id.statusLayout);
        indicatorCircle = findViewById(R.id.indicatorCircle);
        connectedPanel.setVisibility(View.VISIBLE); 
        receivedMsgView = findViewById(R.id.textReceivedMsg);
        receivedMsgView.setText("Received Messages: ");
        connectedText = findViewById(R.id.textConnectedStatus);

        updateConnectionUI(myApp.btConnection() != null, null);

        saberBlade = findViewById(R.id.saberBlade);
    }

    private void animateLightsaber(boolean isExtending) {
        if (saberBlade == null) return;

        if (saberAnimator != null && saberAnimator.isRunning()) {
            saberAnimator.cancel();
        }

        saberBlade.post(() -> {
            View parentContainer = (View) saberBlade.getParent();
            if (parentContainer == null) return;

            int boundWidth = parentContainer.getWidth();
            int hiltSize = findViewById(R.id.saberHandle).getWidth();
            int sidePadding = parentContainer.getPaddingStart() + parentContainer.getPaddingEnd();

            int computedLength = isExtending ? (boundWidth - hiltSize - sidePadding) : 0;
            computedLength = Math.max(computedLength, 0);

            int startWidth = saberBlade.getWidth();
            if (startWidth == computedLength) return;

            saberAnimator = android.animation.ValueAnimator.ofInt(startWidth, computedLength);
            saberAnimator.setDuration(450); 
            saberAnimator.setInterpolator(new android.view.animation.OvershootInterpolator(0.8f)); 
            saberAnimator.addUpdateListener(anim -> {
                int frameValue = (int) anim.getAnimatedValue();
                LinearLayout.LayoutParams lp = (LinearLayout.LayoutParams) saberBlade.getLayoutParams();
                lp.width = frameValue;
                saberBlade.setLayoutParams(lp);
            });
            saberAnimator.start();
        });
    }

    private void updateConnectionUI(boolean isConnected, String deviceName) {
        if (indicatorCircle != null) {
            indicatorCircle.setBackgroundTintList(android.content.res.ColorStateList.valueOf(
                    androidx.core.content.ContextCompat.getColor(this,
                            isConnected ? R.color.status_success : R.color.neon_red)));
        }
        if (connectedText != null) {
            connectedText.setTextColor(androidx.core.content.ContextCompat.getColor(this,
                    isConnected ? R.color.status_success : R.color.neon_red));
            if (isConnected) {
                connectedText.setText(deviceName != null ? "Connected to " + deviceName : "CONNECTED");
            } else {
                connectedText.setText("DISCONNECTED");
            }
        }
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        getApplicationContext().unregisterReceiver(infoReceiver);
        getApplicationContext().unregisterReceiver(msgReceiver);
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions,
            @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestCode == BLUETOOTH_PERMISSIONS_REQUEST_CODE) {
            boolean allPermissionsGranted = true;
            for (int result : grantResults) {
                if (result != PackageManager.PERMISSION_GRANTED) {
                    allPermissionsGranted = false;
                    break;
                }
            }
            if (allPermissionsGranted) {
                
                if (!myApp.btInterface().isBluetoothEnabled()) {
                    Toast.makeText(this, "This app requires Bluetooth to function.", Toast.LENGTH_SHORT).show();
                    Intent enableBtIntent = new Intent(BluetoothAdapter.ACTION_REQUEST_ENABLE);
                    requestEnableBluetooth.launch(enableBtIntent);
                } else {
                    
                    bootBluetoothService();
                }
            } else {
                
                showPermissionDeniedDialog();
            }
        }
    }

    private void enableDeviceDiscovery() {
        Intent discoverableIntent = new Intent(BluetoothAdapter.ACTION_REQUEST_DISCOVERABLE);
        discoverableIntent.putExtra(BluetoothAdapter.EXTRA_DISCOVERABLE_DURATION, DISCOVERABLE_DURATION);
        requestDiscoverable.launch(discoverableIntent);
        myApp.btInterface().acceptIncomingConnection();
    }

    private void reloadDiscoveredDevices() {
        
        bluetoothDeviceAdapter.initPairedDevices(myApp.btInterface().getBondedDevices());
        myApp.btInterface().scanForDevices();
    }

    private void bootBluetoothService() {
        if (myApp.btInterface().isBluetoothEnabled()) {
            bluetoothDeviceAdapter.initPairedDevices(myApp.btInterface().getBondedDevices());
        }
    }

    private void showPermissionDeniedDialog() {
        new AlertDialog.Builder(this)
                .setTitle("Bluetooth Access Revoked")
                .setMessage("Bluetooth and Location permissions are strictly required for establishing connections. Fix this in the App Settings.")
                .setPositiveButton("Goto Settings", (dialog, which) -> {
                    Intent navIntent = new Intent(Settings.ACTION_APPLICATION_DETAILS_SETTINGS);
                    navIntent.setData(Uri.fromParts("package", getPackageName(), null));
                    startActivity(navIntent);
                })
                .setNegativeButton("Ignore", null)
                .show();
    }

    @SuppressLint("MissingPermission")
    private void onBluetoothInfoReceived(Intent bcastIntent, String actionOrigin) {
        Log.d(TAG, "Parsing Broadcast: " + actionOrigin);
        
        if (actionOrigin.equals(BluetoothDevice.ACTION_FOUND)) {
            BluetoothDevice newlyFound = bcastIntent.getParcelableExtra(BluetoothDevice.EXTRA_DEVICE, BluetoothDevice.class);
            if (newlyFound != null) {
                Log.d(TAG, "Spotted BT node -> " + newlyFound.getName() + " [" + newlyFound.getAddress() + "]");
                bluetoothDeviceAdapter.addDiscoveredDevice(newlyFound);
            }
        } else if (actionOrigin.equals(BluetoothConnection.ACTION_CONNECTED)) {
            boolean isLinkActive = bcastIntent.getBooleanExtra(BluetoothConnection.EXTRA_CONNECTED, false);
            BluetoothDevice peerDevice = bcastIntent.getParcelableExtra(BluetoothConnection.EXTRA_DEVICE, BluetoothDevice.class);
            
            if (peerDevice != null) {
                Log.i(TAG, "Link established with " + peerDevice.getName() + ": " + isLinkActive);
                if (!isLinkActive) {
                    Toast.makeText(this, "Dropped Connection... Re-arming in 3s...", Toast.LENGTH_SHORT).show();
                    new android.os.Handler(android.os.Looper.getMainLooper()).postDelayed(() -> {
                        myApp.btInterface().connectAsClient(peerDevice);
                    }, 3000);
                } else {
                    if (discoverSwitch != null && discoverSwitch.isChecked()) {
                        discoverSwitch.setChecked(false);
                        animateLightsaber(false);
                    }
                }
                updateConnectionUI(isLinkActive, peerDevice.getName());
                reloadDiscoveredDevices();
            }
        }
    }

    private void onMsgReceived(BluetoothMessage btMsg) {
        if (btMsg instanceof BluetoothMessage.PlainStringMessage m) {
            receivedMsgView.append(m.rawMsg() + "\n");
        } else if (btMsg instanceof BluetoothMessage.TargetFoundMessage m) {
            receivedMsgView.append("[image-rec] " + m.rawMsg() + "\n"); 
        } else if (btMsg instanceof BluetoothMessage.RobotPositionMessage m) {
            receivedMsgView.append("[location] " + m.rawMsg() + "\n"); 
        }
    }

    private void toggleTheme() {
        android.content.SharedPreferences prefs = getSharedPreferences("AppPrefs", MODE_PRIVATE);
        boolean currentMode = prefs.getBoolean("DarkMode", false);
        boolean newMode = !currentMode;

        prefs.edit().putBoolean("DarkMode", newMode).apply();

        androidx.appcompat.app.AppCompatDelegate.setDefaultNightMode(
                newMode ? androidx.appcompat.app.AppCompatDelegate.MODE_NIGHT_YES
                        : androidx.appcompat.app.AppCompatDelegate.MODE_NIGHT_NO);
    }
}

File: android_app/app/src/main/java/com/example/android_app/bluetooth/BluetoothActivity.kt
package com.example.android_app.bluetooth

import android.Manifest
import android.annotation.SuppressLint
import android.bluetooth.BluetoothAdapter
import android.bluetooth.BluetoothDevice
import android.bluetooth.BluetoothManager
import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.content.pm.PackageManager
import android.os.Build
import android.os.Bundle
import android.widget.Button
import android.widget.EditText
import android.widget.TextView
import android.widget.Toast
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import androidx.lifecycle.Lifecycle
import androidx.lifecycle.lifecycleScope
import androidx.lifecycle.repeatOnLifecycle
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.example.android_app.R
import kotlinx.coroutines.flow.launchIn
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.launch

class BluetoothActivity : AppCompatActivity() {

    private lateinit var pairedAdapter: DeviceListAdapter
    private lateinit var discoveredAdapter: DeviceListAdapter
    private lateinit var statusView: TextView
    private lateinit var inboundLog: TextView
    private lateinit var sendField: EditText

    private val bluetoothAdapter: BluetoothAdapter? by lazy {
        (getSystemService(Context.BLUETOOTH_SERVICE) as? BluetoothManager)?.adapter
    }

    private val discoveryReceiver = object : BroadcastReceiver() {
        override fun onReceive(context: Context?, intent: Intent?) {
            when (intent?.action) {
                BluetoothDevice.ACTION_FOUND -> {
                    val device: BluetoothDevice? =
                        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.TIRAMISU) {
                            intent.getParcelableExtra(
                                BluetoothDevice.EXTRA_DEVICE,
                                BluetoothDevice::class.java,
                            )
                        } else {
                            @Suppress("DEPRECATION")
                            intent.getParcelableExtra(BluetoothDevice.EXTRA_DEVICE)
                        }
                    device?.let { discoveredAdapter.addIfMissing(it) }
                }
                BluetoothAdapter.ACTION_DISCOVERY_FINISHED -> {
                    findViewById<Button>(R.id.btnScan).isEnabled = true
                }
            }
        }
    }

    private val enableBtLauncher =
        registerForActivityResult(ActivityResultContracts.StartActivityForResult()) {
            /* nothing — we re-check on scan button press */
        }

    private enum class PendingAction { NONE, LOAD_PAIRED, START_SCAN }
    private var pendingAction: PendingAction = PendingAction.NONE

    private val permissionsLauncher =
        registerForActivityResult(ActivityResultContracts.RequestMultiplePermissions()) { grants ->
            val allGranted = grants.values.all { it }
            val action = pendingAction
            pendingAction = PendingAction.NONE
            if (!allGranted) {
                toast("Bluetooth permissions required")
                return@registerForActivityResult
            }
            when (action) {
                PendingAction.LOAD_PAIRED -> loadPairedDevices()
                PendingAction.START_SCAN -> startDiscovery()
                PendingAction.NONE -> Unit
            }
        }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_bluetooth)

        statusView = findViewById(R.id.statusView)
        inboundLog = findViewById(R.id.inboundLog)
        sendField = findViewById(R.id.sendField)

        pairedAdapter = DeviceListAdapter(::onDeviceSelected)
        discoveredAdapter = DeviceListAdapter(::onDeviceSelected)

        findViewById<RecyclerView>(R.id.pairedList).apply {
            layoutManager = LinearLayoutManager(this@BluetoothActivity)
            adapter = pairedAdapter
        }
        findViewById<RecyclerView>(R.id.discoveredList).apply {
            layoutManager = LinearLayoutManager(this@BluetoothActivity)
            adapter = discoveredAdapter
        }

        findViewById<Button>(R.id.btnScan).setOnClickListener { onScanClicked() }
        findViewById<Button>(R.id.btnDisconnect).setOnClickListener {
            BluetoothService.disconnect()
        }
        findViewById<Button>(R.id.btnStopConnection).setOnClickListener {
            BluetoothService.stopConnection()
        }
        findViewById<Button>(R.id.btnSend).setOnClickListener {
            val text = sendField.text.toString().trim()
            if (text.isEmpty()) return@setOnClickListener
            if (BluetoothService.send(text)) {
                sendField.text.clear()
            } else {
                toast("Not connected")
            }
        }

        registerReceiver(discoveryReceiver, IntentFilter().apply {
            addAction(BluetoothDevice.ACTION_FOUND)
            addAction(BluetoothAdapter.ACTION_DISCOVERY_FINISHED)
        })

        observeService()
        loadPairedDevices()
    }

    override fun onDestroy() {
        super.onDestroy()
        try {
            unregisterReceiver(discoveryReceiver)
        } catch (_: IllegalArgumentException) {}
        cancelDiscovery()
    }

    private fun observeService() {
        lifecycleScope.launch {
            repeatOnLifecycle(Lifecycle.State.STARTED) {
                launch {
                    BluetoothService.state.onEach { state ->
                        val name = BluetoothService.connectedDevice.value
                        statusView.text = when (state) {
                            BluetoothService.State.DISCONNECTED ->
                                getString(R.string.status_disconnected)
                            BluetoothService.State.CONNECTING ->
                                getString(R.string.status_connecting)
                            BluetoothService.State.CONNECTED ->
                                getString(R.string.status_connected, name ?: "device")
                            BluetoothService.State.RECONNECTING ->
                                getString(R.string.status_reconnecting)
                        }
                    }.launchIn(this)
                }
                launch {
                    BluetoothService.messages.onEach { line ->
                        appendInbound(line)
                    }.launchIn(this)
                }
            }
        }
    }

    private fun appendInbound(line: String) {
        val current = inboundLog.text?.toString().orEmpty()
        val trimmed = if (current.length > 4000) current.takeLast(2000) else current
        inboundLog.text = if (trimmed.isEmpty()) line else "$trimmed\n$line"
    }

    private fun onDeviceSelected(device: BluetoothDevice) {
        cancelDiscovery()
        BluetoothService.connect(device)
    }

    private fun onScanClicked() {
        if (!requireBtEnabled()) return
        if (!hasScanPermissions()) {
            pendingAction = PendingAction.START_SCAN
            permissionsLauncher.launch(neededScanPermissions())
            return
        }
        startDiscovery()
    }

    @SuppressLint("MissingPermission")
    private fun startDiscovery() {
        val adapter = bluetoothAdapter ?: run {
            toast("No Bluetooth adapter")
            return
        }
        discoveredAdapter.submit(emptyList())
        findViewById<Button>(R.id.btnScan).isEnabled = false
        adapter.cancelDiscovery()
        adapter.startDiscovery()
    }

    @SuppressLint("MissingPermission")
    private fun cancelDiscovery() {
        bluetoothAdapter?.cancelDiscovery()
    }

    @SuppressLint("MissingPermission")
    private fun loadPairedDevices() {
        if (!hasConnectPermission()) {
            pendingAction = PendingAction.LOAD_PAIRED
            permissionsLauncher.launch(neededScanPermissions())
            return
        }
        val adapter = bluetoothAdapter ?: return
        pairedAdapter.submit(adapter.bondedDevices.toList())
    }

    private fun requireBtEnabled(): Boolean {
        val adapter = bluetoothAdapter ?: run {
            toast("No Bluetooth adapter")
            return false
        }
        if (!adapter.isEnabled) {
            enableBtLauncher.launch(Intent(BluetoothAdapter.ACTION_REQUEST_ENABLE))
            return false
        }
        return true
    }

    private fun hasScanPermissions(): Boolean = neededScanPermissions().all { perm ->
        ContextCompat.checkSelfPermission(this, perm) == PackageManager.PERMISSION_GRANTED
    }

    private fun hasConnectPermission(): Boolean =
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            ContextCompat.checkSelfPermission(this, Manifest.permission.BLUETOOTH_CONNECT) ==
                    PackageManager.PERMISSION_GRANTED
        } else true

    private fun neededScanPermissions(): Array<String> =
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.S) {
            arrayOf(
                Manifest.permission.BLUETOOTH_SCAN,
                Manifest.permission.BLUETOOTH_CONNECT,
            )
        } else {
            arrayOf(
                Manifest.permission.ACCESS_FINE_LOCATION,
                Manifest.permission.ACCESS_COARSE_LOCATION,
            )
        }

    private fun toast(msg: String) {
        Toast.makeText(this, msg, Toast.LENGTH_SHORT).show()
    }
}